Fix the mid-level pass pipeline

atrick opened the pull request:
This is a small series of commits that together fixes issues in the pass pipeline that are causing severe instability in benchmark performance and blocking other bug fixes.

We have heroic optimizations for array access, but whether those actually kick in was essentially random. Now, those optimizations always have a chance to run and the pass pipeline consistently reruns when necessary as it was designed to do.

The two main bug fixes are
- Don't inline array semantic calls before the mid-level pipeline
- Fix the mid-level pipeline restart mechanism

atrick committed:
Add -sil-inline-never-functions flag.

And fix a test case that assumes Array.append is never inlined.

We need to be able to prevent inlining a function and any of its
specializations. There's no way to specify multiple function names on
the command line, so we use the partial match technique.

atrick committed:
Disable early inlining of array.append.

Inlining semantic calls must *always* be deferred to the mid-level
pipeline. Otherwise, array optimizations take effect randomly. This is
a primary cause of instability in Swift benchmarks.

atrick committed:
Add a test case for the interplay between array optimization and
inlining.

This tests that the pass pipeline is properly setup to run all the
array optimizations even when inlining is required. This requires that
the passmanager defer all semantic inlining until it reaches the
mid-level pipeline.

atrick committed:
Move the FSO bailout to the beginning of the pass.

This way we catch the DeadFunctionArgs optimization too, which is a
separate sub-pass.

atrick commented:
This PR is on hold because there are a few benchmark regressions I don't have time to investigate today. The vast majority are improvements, but I should still understand why the few regressions happened before merging.

atrick committed:
Emit inlining diagnostics as debug messages.

So -debug-only=sil-inliner is useful.

atrick committed:
Add an inliner heuristics for semantic call sites.

This complements the pass pipeline changes. Now no semantic inlining
occurs in the early phase, and we expect to see nested semantic calls
in the mid-level pipeline. To ensure that optimizations based on call
semantics work, we need those nested semantic calls to be inlined.

atrick committed:
Don't inline semantics calls into generic functions.

atrick committed:
Fix the mid-level pass pipeline.

Module passes need to be in a separate pipeline, otherwise the
pipeline restart mechanism will be broken.

This makes GlobalOpt and serialization run earlier in the
pipeline. There's no explicit reason for them to be run later, in the
middle of a function pass pipeline.

Also, pipeline boundaries, like serialization and module passes should
be explicit at the the top level function that creates the pass
pipelines.

atrick commented:
This PR is still active. I plan to work on merging it this week after finishing some work on EscapeAnalysis. Unfortunately, I can't avoid all regressions because of some fundamental problems in ARC optimization.

eeckstein commented on test/SILOptimizer/array_contentof_opt.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-swift-frontend -O -sil-verify-all -emit-sil -enforce-exclusivity=unchecked  %s | %FileCheck %s
+// RUN: %target-swift-frontend -O -sil-verify-all -emit-sil -enforce-exclusivity=unchecked -Xllvm '-sil-inline-never-functions=$sSa6appendyyxn' %s | %FileCheck %s
```
The purpose of this test is to check if the array.append(contentof) optimization works.
With your change it does not work anymore. Adding this "workaround" option to let the test pass does not make any sense.

atrick replied:
@eeckstein 
As the commit log said:
fix a test case that assumes Array.append is never inlined.

I just added another comment to make it more clear:
// The checks lines are written assuming that the array(Element) calls
// are never inlined. Consequently, the RUN line must prevent inlining
// Array(Element).

What makes you think this PR broke an optimization?

eeckstein replied:
As written in the comment in the test file: This is an end-to-end test of the array(contentsOf) -> array(Element) optimization

This optimization replaces an Array.append(contentof:) with a sequence of Array.append(element).
The point of this test is to check if the optimization works. And working means: out of the box without any special command line options.

atrick replied:
@eeckstein I don't understand why you think the optimization is broken. If array(Element) is inlined after the optimization, then the test's CHECK lines will fail. So it wouldn't be an effective end-to-end test. A proper end-to-end test is going to need to control inlining if it's checking for the presence of calls.

eeckstein replied:
Sorry, I misunderstood the change here. Looks good!

atrick replied:
I cleared up the test case comments
https://github.com/apple/swift/pull/22445/commits/1e73e548caabf8bcf8d598105adadaea67f4d594

Thanks

eeckstein reviewed (approved):
lgtm

atrick committed:
Fix generic specialization test case.

atrick committed:
Add @inline annotations to merge_exclusivity test to make it stable.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci test source compatibility

atrick commented:
@swift-ci test compiler performance

atrick commented:
@swift-ci test compiler performance

atrick commented:
@swift-ci test compiler performance

futurejones commented:
Can one of the admins verify this patch?

atrick committed:
Add PointerLikeTypeTraits to ApplySite and FullApplySite.

So they can be used in ADTs like SmallPtrSet.

atrick committed:
Add getArraySemanticsKind API.

So that it's possible to determine the kind of Array semantic function
from its SILFunction decl only.

atrick committed:
SILOptimizer: Add enforcement of function-pass pipelines.

Don't allow module passes to be inserted within a function pass
pipeline. This silently breaks the function pipeline both interfering
with analysis and the normal pipeline restart mechanism.

atrick committed:
SILOptimizer: Fix analysis invalidation after devirtualization.

Devirtualizing try_apply modified the CFG, but passes that run
devirtualization were not invalidating any CFG analyses, such as the
domtree.

This could hypothetically have caused miscompilation, but will be
caught by running with -sil-verify-all.

atrick committed:
WIP: Fix SIL pipeline and inlined to preserve @_semantic functions.

Run all high-level opts before inlining any semantic calls. We have
one exception: an array.uninitialized compiler intrinsic that needs to
be expanded early for optimizations to see the "normal" underlying
semantic call.

This not only ensures that stdlib is well optimized, but ensures that
callers are fully optimized and inlined before attempting to run
semantic passes.

Lower nested semantic calls incrementally so that semantic passes
rerun at each level.

Avoid fully expanding any semantic functions until all semantic passes
have seen them and all callers have been fully optimized.

Inlining the lowest-level semantic calls defeats basic EscapeAnalysis
and SideEffectAnalysis. This shouldn't be done until the late inliner
runs.

atrick committed:
SIL pass pipeline comments and minor adjustment.

atrick committed:
Add @_semantics to intermediate Array methods.

Interposing non-semantic calls within semantic calls breaks the
mechanism by which the pass pipeline incrementally processes each
layer of semantics.

Semantics should go "all the way down" to the level that we want
inlined in the mid-level pipeline.

atrick committed:
Adjust a test case for @_semantic call processing.

Which broke as a result of stdlib refactoring, and which I don't have
time to get working again.

atrick committed:
WIP: Fix SimplifyCFG analysis invalidation.

Not all simplifications correctly invalidate CFG analyses.

atrick commented:
Superseded by #31163 and other followup PRs.
